File: .github/scripts/k0s-cni-migration-test.py
#!/usr/bin/env python3
"""Behavior regression for k0s CNI migration after controller rewrite."""

import json
import os
import pathlib
import subprocess
import tempfile
import textwrap


ROOT = pathlib.Path(__file__).resolve().parents[2]
TASKFILE = ROOT / "infra/tasks/k0s.yml"
START = '        cni_migrated="$cni_was_unmasqueraded"\n'
END = (
    "        if sudo -n test -e "
    "/var/lib/hive/k0s-recovery/neo4j-secrets.yaml.enc; then\n"
)


def migration_source():
    taskfile = TASKFILE.read_text(encoding="utf-8")
    embedded = taskfile.split(START, 1)[1].split(END, 1)[0]
    return 'cni_migrated="$cni_was_unmasqueraded"\n' + textwrap.dedent(embedded)


def main():
    with tempfile.TemporaryDirectory() as directory:
        work = pathlib.Path(directory)
        mock_bin = work / "bin"
        mock_bin.mkdir()
        log = work / "commands.log"
        namespace_state = work / "namespace-state.json"
        namespace_state.write_text(
            json.dumps({"hive-data": "data", "hive-system": "workers"}),
            encoding="utf-8",
        )
        namespace_manifest = work / "infra/k0s/manifests/namespaces.yaml"
        namespace_manifest.parent.mkdir(parents=True)
        namespace_manifest.write_text(
            (ROOT / "infra/k0s/manifests/namespaces.yaml").read_text(
                encoding="utf-8"
            ),
            encoding="utf-8",
        )
        cni = work / "10-kuberouter.conflist"
        cni.write_text(
            '{"plugins":[{"type":"bridge","ipMasq":true}]}\n',
            encoding="utf-8",
        )
        sudo = mock_bin / "sudo"
        sudo.write_text(
            textwrap.dedent(
                """\
                #!/usr/bin/env python3
                import json
                import os
                import pathlib
                import subprocess
                import sys

                args = sys.argv[1:]
                if args and args[0] == "-n":
                    args = args[1:]
                with pathlib.Path(os.environ["MOCK_LOG"]).open("a") as log:
                    log.write(" ".join(args) + "\\n")
                if args[:2] == ["test", "-s"]:
                    raise SystemExit(0 if pathlib.Path(args[2]).stat().st_size else 1)
                if args and args[0] == "jq":
                    raise SystemExit(subprocess.run(args).returncode)
                if args[:4] == ["k0s", "kubectl", "create", "namespace"]:
                    print("apiVersion: v1")
                    print("kind: Namespace")
                    print("metadata:")
                    print(f"  name: {args[4]}")
                if args[:3] == ["k0s", "kubectl", "apply"]:
                    if args[-2:] == ["-f", "-"]:
                        manifest = sys.stdin.read()
                    else:
                        manifest = pathlib.Path(args[-1]).read_text()
                    state_path = pathlib.Path(os.environ["MOCK_NAMESPACE_STATE"])
                    state = json.loads(state_path.read_text())
                    for document in manifest.split("---"):
                        if "kind: Namespace" not in document:
                            continue
                        namespace = next(
                            (
                                line.split(":", 1)[1].strip()
                                for line in document.splitlines()
                                if line.strip().startswith("name:")
                            ),
                            "",
                        )
                        if namespace not in state:
                            continue
                        role = next(
                            (
                                line.split(":", 1)[1].strip()
                                for line in document.splitlines()
                                if line.strip().startswith("hive.nook.sh/role:")
                            ),
                            "",
                        )
                        state[namespace] = role
                    state_path.write_text(json.dumps(state))
                if args[:4] == ["k0s", "kubectl", "get", "deployment/hive"]:
                    raise SystemExit(0)
                if args[:4] == [
                    "k0s",
                    "kubectl",
                    "get",
                    "deployment/hive-workbench-dispatcher",
                ]:
                    raise SystemExit(0)
                if args[:4] == [
                    "k0s",
                    "kubectl",
                    "get",
                    "deployment/hive-reaper-controller",
                ]:
                    raise SystemExit(0)
                """
            ),
            encoding="utf-8",
        )
        sudo.chmod(0o755)
        harness = work / "harness.sh"
        harness.write_text(
            f"""#!/usr/bin/env bash
set -euo pipefail
cni_config={cni}
cni_config_next=""
cni_was_unmasqueraded=true
remote_dir={work}
{migration_source()}
""",
            encoding="utf-8",
        )
        harness.chmod(0o755)
        env = {
            **os.environ,
            "PATH": f"{mock_bin}:{os.environ['PATH']}",
            "MOCK_LOG": str(log),
            "MOCK_NAMESPACE_STATE": str(namespace_state),
        }
        subprocess.run([str(harness)], env=env, check=True)
        assert json.loads(namespace_state.read_text(encoding="utf-8")) == {
            "hive-data": "data",
            "hive-system": "workers",
        }
        commands = log.read_text(encoding="utf-8")
        namespace_apply = (
            "k0s kubectl apply "
            f"-f {namespace_manifest}"
        )
        assert namespace_apply in commands, namespace_apply
        for deployment in (
            "hive",
            "hive-workbench-dispatcher",
            "hive-reaper-controller",
        ):
            restart = (
                "k0s kubectl rollout restart "
                f"deployment/{deployment} --namespace hive-system"
            )
            status = (
                "k0s kubectl rollout status "
                f"deployment/{deployment} --namespace hive-system --timeout=10m"
            )
            assert restart in commands, restart
            assert status in commands, status
        assert "k0s kubectl rollout restart deployment/coredns" in commands
        print("k0s CNI rewrite migration rollouts: ok")


if __name__ == "__main__":
    main()

File: .github/scripts/k0s-cni-migration-test.ts
import { err, ok, type Result } from "neverthrow";
import {
  FixtureFile,
  FixtureDirectory,
  FixtureWorkspaceRequest,
  FixtureEmbeddedSource,
} from "./operational-fixture";
import {
  OperationalCommandProbe,
  OperationalProbeStream,
  OperationalContractFailureKind,
  type OperationalContractFailure,
} from "./operational-contract";
import { join, resolve } from "node:path";

const root = resolve(import.meta.dir, "../..");
const { PATH: executablePath = "" } = process.env;
const taskfile = resolve(root, "infra/tasks/k0s.yml");
const start = '        cni_migrated="$cni_was_unmasqueraded"\n';
const end =
  "        if sudo -n test -e /var/lib/hive/k0s-recovery/neo4j-secrets.yaml.enc; then\n";

class CniMigrationSource {
  read(): Result<string, OperationalContractFailure> {
    const task = new FixtureFile(taskfile).read();
    if (task.isErr()) return err(task.error);
    return new FixtureEmbeddedSource(task.value)
      .between(start, end)
      .map((embedded) => start.trimStart() + embedded.replace(/^ {8}/gm, ""));
  }
}
class CniMigrationFixture {
  constructor(private readonly work: string) {}
  execute(): Result<void, OperationalContractFailure> {
    const work = this.work;
    const mockBin = join(work, "bin");
    const log = join(work, "commands.log");
    const namespaceManifest = join(work, "infra/k0s/manifests/namespaces.yaml");
    const prepared1 = new FixtureDirectory(mockBin).create();
    if (prepared1.isErr()) return err(prepared1.error);
    const prepared2 = new FixtureDirectory(
      resolve(namespaceManifest, ".."),
    ).create();
    if (prepared2.isErr()) return err(prepared2.error);
    const prepared6 = new FixtureFile(namespaceManifest).copyFrom(
      resolve(root, "infra/k0s/manifests/namespaces.yaml"),
    );
    if (prepared6.isErr()) return err(prepared6.error);
    const manifestResult = new FixtureFile(namespaceManifest).read();
    if (manifestResult.isErr()) return err(manifestResult.error);
    const manifest = manifestResult.value;
    for (const fragment of [
      "name: hive-data",
      "hive.nook.sh/role: data",
      "name: hive-system",
      "hive.nook.sh/role: workers",
    ]) {
      if (!manifest.includes(fragment))
        return err({
          kind: OperationalContractFailureKind.Requirement,
          message: `namespace manifest missing: ${fragment}`,
        });
    }
    const cni = join(work, "10-kuberouter.conflist");
    const prepared3 = new FixtureFile(cni).write(
      '{"plugins":[{"type":"bridge","ipMasq":true}]}\n',
    );
    if (prepared3.isErr()) return err(prepared3.error);
    const fixtureCommand = join(mockBin, "fixture-command");
    const prepared4 = new FixtureFile(fixtureCommand).write(
      `#!/usr/bin/env bash
set -euo pipefail
if test "\${1:-}" = -n; then shift; fi
printf '%s\\n' "$*" >> "$MOCK_LOG"
case "\${1:-}" in
  test)
    if test "$2" = -s; then test -s "$3"; else exit 1; fi
    ;;
  jq)
    exec jq "\${@:2}"
    ;;
  install)
    cp "\${@: -2:1}" "\${@: -1}"
    ;;
  k0s)
    exit 0
    ;;
  *)
    printf 'unexpected sudo command: %s\\n' "$*" >&2
    exit 2
    ;;
esac
`,
    );
    if (prepared4.isErr()) return err(prepared4.error);
    const prepared7 = new FixtureFile(fixtureCommand).makeExecutable();
    if (prepared7.isErr()) return err(prepared7.error);
    const harness = join(work, "harness.sh");
    const migration = new CniMigrationSource().read();
    if (migration.isErr()) return err(migration.error);
    const fixtureMigration = migration.value.replaceAll(
      "sudo -n ",
      `${JSON.stringify(fixtureCommand)} `,
    );
    const prepared5 = new FixtureFile(harness).write(
      `#!/usr/bin/env bash
set -euo pipefail
export MOCK_LOG=${JSON.stringify(log)}
cni_config=${cni}
cni_config_next=""
cni_was_unmasqueraded=true
remote_dir=${work}
${fixtureMigration}
`,
    );
    if (prepared5.isErr()) return err(prepared5.error);
    const prepared8 = new FixtureFile(harness).makeExecutable();
    if (prepared8.isErr()) return err(prepared8.error);
    const runInput = {
      cmd: [harness],
      env: {
        ...process.env,
        PATH: `${mockBin}:${executablePath}`,
        MOCK_LOG: log,
      },
      stdout: OperationalProbeStream.Inherit,
      stderr: OperationalProbeStream.Inherit,
    };
    const run = new OperationalCommandProbe(runInput).execute();
    if (run.isErr()) return err(run.error);
    const result = run.value;
    if (result.exitCode !== 0)
      return err({
        kind: OperationalContractFailureKind.Requirement,
        message: `CNI migration harness exited ${result.exitCode}`,
      });
    const commandsResult = new FixtureFile(log).read();
    if (commandsResult.isErr()) return err(commandsResult.error);
    const commands = commandsResult.value;
    const namespaceApply = `k0s kubectl apply -f ${namespaceManifest}`;
    if (!commands.includes(namespaceApply))
      return err({
        kind: OperationalContractFailureKind.Requirement,
        message: `missing command: ${namespaceApply}`,
      });
    for (const deployment of [
      "hive",
      "hive-workbench-dispatcher",
      "hive-reaper-controller",
    ]) {
      for (const command of [
        `k0s kubectl rollout restart deployment/${deployment} --namespace hive-system`,
        `k0s kubectl rollout status deployment/${deployment} --namespace hive-system --timeout=10m`,
      ]) {
        if (!commands.includes(command))
          return err({
            kind: OperationalContractFailureKind.Requirement,
            message: `missing command: ${command}`,
          });
      }
    }
    if (!commands.includes("k0s kubectl rollout restart deployment/coredns")) {
      return err({
        kind: OperationalContractFailureKind.Requirement,
        message: "CNI migration must restart CoreDNS",
      });
    }
    console.log("k0s CNI rewrite migration rollouts: ok");
    return ok();
  }
}
class CniMigrationContract {
  execute(): Result<void, OperationalContractFailure> {
    const workspace = new FixtureWorkspaceRequest(
      "nook-cni-migration-",
    ).create();
    if (workspace.isErr()) return err(workspace.error);
    return workspace.value.finish(
      new CniMigrationFixture(workspace.value.path).execute(),
    );
  }
}
const outcome = new CniMigrationContract().execute();
if (outcome.isErr()) {
  console.error(outcome.error.message);
  process.exitCode = 1;
}
